Sources/NIOHTTP1/HTTPDecoder.swift

```diff
@@ -357,92 +377,174 @@ public class HTTPDecoder<HTTPMessageT>: ByteToMessageDecoder, AnyHTTPDecoder {
+        }
     }
 
-    public func decode(ctx: ChannelHandlerContext, buffer: inout ByteBuffer) throws -> DecodingState {
-        if let slice = state.slice {
-            // If we stored a slice before we need to ensure we move the readerIndex so we don't try to parse the data again. We
-            // also need to update the reader index to whatever it is now.
-            state.slice = (buffer.readerIndex, slice.length)
-            buffer.moveReaderIndex(forwardBy: state.readerIndexAdjustment)
+    // Decode HTTP until there is nothing more to decode.
+    private func decodeHTTP(ctx: ChannelHandlerContext) throws {
+        // We need to refetch the cumulationBuffer on each loop as it may has changed due re-entrance calls of channelRead(...)
+        while let bufferSlice = self.cumulationBuffer, bufferSlice.readableBytes > 0 {
+            // Using withVeryUnsafeBytes here as this simplifies the calculation of the readerIndex which is relative to the baseAddress.
+            let result = bufferSlice.withVeryUnsafeBytes { (pointer) -> size_t in
+                self.state.baseAddress = pointer.baseAddress!.assumingMemoryBound(to: UInt8.self)
+                defer {
+                    self.state.baseAddress = nil
+                }
+                return self.state.baseAddress!.withMemoryRebound(to: Int8.self, capacity: pointer.count) { p in
+                    c_nio_http_parser_execute(&self.parser, &self.settings, p.advanced(by: bufferSlice.readerIndex), bufferSlice.readableBytes)
+                }
+            }
+            try self.rethrowParserError()
+
+            /// http_parser_execute(...) should always consume the whole buffer.
+            assert(result == bufferSlice.readableBytes)
+
+            // Update readerIndex of the cumulationBuffer itself as we will refetch it in the next loop run if needed.
+            self.cumulationBuffer?.moveReaderIndex(forwardBy: result)
+        }
+
+        if self.state.seenEOF {
+            // We need to notify the parser about the EOF as we received it while in http_parser_execute.
+            self.notifyParserEOF(ctx: ctx)
         }
+    }
+
+    private func discardDecodedBytes() {
+        guard self.cumulationBuffer != nil else {
+            // Guard against the case of closing the channel. In this case the cumulationBuffer will be nil.
+            return
+        }
+
+        assert(self.cumulationBuffer!.readableBytes == 0)
+
+        switch self.state.dataAwaitingState {
+        case .body, .messageBegin:
+            assert(self.state.currentNameIndex == nil)
+            assert(self.state.currentHeaders.isEmpty)
+            assert(self.state.slice == nil)
+
+            // Its safe to just drop the cumulationBuffer as we not have any extra views into it that are represented as readerIndex / length.
+            self.cumulationBuffer = nil
+
+        case .headerField, .headerValue:
+            if let headerStartIdx = self.state.headerStartIndex {
```

Comment:
The whole body of this case is in the `if` block, so this may as well be `guard`.